deepspeed/inference/engine.py

```diff
@@ -480,25 +494,34 @@ def _create_cuda_graph(self, *inputs, **kwargs):
                 ret = self.module(*inputs, **kwargs)
         get_accelerator().current_stream().wait_stream(cuda_stream)
 
+        shape_signature = self._input_signature(*inputs, **kwargs)
         # create cuda_graph and assign static_inputs and static_outputs
-        self._cuda_graphs = torch.cuda.CUDAGraph()
-        self.static_inputs = inputs
-        self.static_kwargs = kwargs
+        self._cuda_graphs[shape_signature] = torch.cuda.CUDAGraph()
+        self._static_inputs[shape_signature] = inputs
+        self._static_kwargs[shape_signature] = kwargs
 
-        with torch.cuda.graph(self._cuda_graphs):
-            self.static_output = self.module(*self.static_inputs, **self.static_kwargs)
+        static_inputs = self._static_inputs[shape_signature]
+        static_kwargs = self._static_kwargs[shape_signature]
 
-        self.cuda_graph_created = True
+        with torch.cuda.graph(self._cuda_graphs[shape_signature]):
+            self._static_output[shape_signature] = self.module(
+                *static_inputs,
+                **static_kwargs)
 
     def _graph_replay(self, *inputs, **kwargs):
+        shape_signature = self._input_signature(*inputs, **kwargs)
+        static_inputs = self._static_inputs[shape_signature]
+        static_kwargs = self._static_kwargs[shape_signature]
+        static_output = self._static_output[shape_signature]
+
         for i in range(len(inputs)):
             if torch.is_tensor(inputs[i]):
-                self.static_inputs[i].copy_(inputs[i])
+                static_inputs[i].copy_(inputs[i])
         for k in kwargs:
             if torch.is_tensor(kwargs[k]):
-                self.static_kwargs[k].copy_(kwargs[k])
-        self._cuda_graphs.replay()
-        return self.static_output
+                static_kwargs[k].copy_(kwargs[k])
+        self._cuda_graphs[shape_signature].replay()
```

Comment:
this way of accessing the Graphs can be costly for the performance. instead of accessing the graphs directly from a dictionary, let's use a list of graphs. We either need to have indirect access to the list by having the mapping to the right index of the list for each shape-signature, or get the list-index directly from the shape_signature.

Reply:
the same issue applys for when accessing the static_inputs, static_kwargs, and static_output of that CUDA-Graph (which are lines 513, 514, and 515).